packages/stratimikos/adapters/belos/tpetra/Thyra_BelosTpetraPreconditionerFactory_def.hpp

```diff
@@ -0,0 +1,369 @@
+  TEUCHOS_TEST_FOR_EXCEPT(Teuchos::is_null(fwdOp));
+
+  typedef typename MatrixType::local_ordinal_type local_ordinal_type;
+  typedef typename MatrixType::global_ordinal_type global_ordinal_type;
+  typedef typename MatrixType::node_type node_type;
+
+  typedef Thyra::TpetraLinearOp<scalar_type, local_ordinal_type, global_ordinal_type, node_type> ThyraTpetraLinOp;
+  const Teuchos::RCP<const ThyraTpetraLinOp> thyraTpetraFwdOp = Teuchos::rcp_dynamic_cast<const ThyraTpetraLinOp>(fwdOp);
+  TEUCHOS_TEST_FOR_EXCEPT(Teuchos::is_null(thyraTpetraFwdOp));
+
+  typedef Tpetra::Operator<scalar_type, local_ordinal_type, global_ordinal_type, node_type> TpetraLinOp;
+  const Teuchos::RCP<const TpetraLinOp> tpetraFwdOp = thyraTpetraFwdOp->getConstTpetraOperator();
+  TEUCHOS_TEST_FOR_EXCEPT(Teuchos::is_null(tpetraFwdOp));
+
+  // Belos-specific typedefs:
+  typedef Tpetra::MultiVector<scalar_type, local_ordinal_type, global_ordinal_type, node_type>     TpetraMV; 
+  typedef Belos::TpetraOperator<scalar_type, local_ordinal_type, global_ordinal_type, node_type> BelosTpOp;
+  typedef Belos::LinearProblem<scalar_type, TpetraMV, TpetraLinOp> BelosTpLinProb;
+
+#ifdef THYRA_BELOS_PREC_ENABLE_HALF_PRECISION
+  // CAG: There is nothing special about the combination double-float,
+  //      except that I feel somewhat confident that Trilinos builds
+  //      with both scalar types.
+  typedef typename Teuchos::ScalarTraits<scalar_type>::halfPrecision half_scalar_type;
+  typedef Tpetra::Operator<half_scalar_type, local_ordinal_type, global_ordinal_type, node_type> TpetraLinOpHalf;
+  typedef Tpetra::MultiVector<half_scalar_type, local_ordinal_type, global_ordinal_type, node_type>     TpetraMVHalf; 
+  typedef Belos::TpetraOperator<half_scalar_type, local_ordinal_type, global_ordinal_type, node_type> BelosTpOpHalf;
+  typedef Belos::LinearProblem<half_scalar_type, TpetraMVHalf, TpetraLinOpHalf> BelosTpLinProbHalf;
+#endif
+
+  // Retrieve concrete preconditioner object
+
+  const Teuchos::Ptr<DefaultPreconditioner<scalar_type> > defaultPrec =
+    Teuchos::ptr(dynamic_cast<DefaultPreconditioner<scalar_type> *>(prec));
+  TEUCHOS_TEST_FOR_EXCEPT(Teuchos::is_null(defaultPrec));
+
+  // This check needed to address Issue #535. 
+  Teuchos::RCP<const Teuchos::ParameterList> innerParamList;
+  if (paramList_.is_null ()) {
+    innerParamList = Teuchos::rcp(new Teuchos::ParameterList(*getValidParameters()));
+  }
+  else {
+    innerParamList = paramList_;
+  }
+
+  bool useHalfPrecision = false;
+  if (innerParamList->isParameter("half precision"))
+    useHalfPrecision = innerParamList->get<bool>("half precision");
+
+  const std::string solverType = Teuchos::getParameter<std::string>(*innerParamList, "BelosPrec Solver Type");
+  const Teuchos::RCP<Teuchos::ParameterList> packageParamList = 
+          Teuchos::rcp(new Teuchos::ParameterList(*Teuchos::sublist(innerParamList, "BelosPrec Solver Params")));
```

Comment:
Are you trying to create a copy of the sublist?  I don't see any reason why a copy should be created here.

If you are not trying to create a copy, just use:

```
const auto packageParamList = sublist(innerParamList, "BelosPrec Solver Params");
```

That idiom is that we want the input PL to get updated with default values if they are not already set.

Reply:
In the analogous Ifpack2 adapter (/ifpack2/adapters/thyra/Thyra_Ifpack2PreconditionerFactory_def.hpp) we definitely need a copy of that param list due to a potential for changes down the line.  The sublist function returns a const list. 

Right out of the box, the const sublist doesn't work for the BelosPrec adapter, either. I get a compiler error from the const-ness.  The first layer is in the BelosTpetraOperator class.  But even if I change that class to take a const param list, then it fails when I try to pass that list to SolverFactory.create, which also expects a non-cost list. (And then who knows what else down the line.)

I will leave this paramList copy as-is for now.

Reply:
@jennloe, you are supposed to write the default values of parameters not already set into the user's PL.  See:

* [`virtual void Teuchos::ParameterListAcceptor::setParameterList(const RCP< ParameterList > & 	paramList)`](https://docs.trilinos.org/dev/packages/teuchos/doc/html/classTeuchos_1_1ParameterListAcceptor.html#a8ff906ce725b1759c4c68dc9d4457f8d)

* [`void Teuchos::ParameterList::validateParametersAndSetDefaults(ParameterList const &validParamList, int const 	depth)`](https://docs.trilinos.org/dev/packages/teuchos/doc/html/classTeuchos_1_1ParameterList.html#a3408d34b88539d6f6024faa05856b91a)

I think there is a documentation problem here.

That is why the the above function `Teuchos::ParameterListAcceptor::setParameterList()` takes as **non-const** `ParameterList ` object.

Reply:
Okay, @cgcgcg helped me clear up some confusion with the constness.  I think we've addressed the original issue.  Pushing updated version shortly.